server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/node/FixedLengthStarTreeNode.java

```diff
@@ -0,0 +1,186 @@
+
+    private final int nodeId;
+    private final int firstChildId;
+
+    RandomAccessInput in;
+
+    public FixedLengthStarTreeNode(RandomAccessInput in, int nodeId) throws IOException {
+        this.in = in;
+        this.nodeId = nodeId;
+        firstChildId = getInt(FIRST_CHILD_ID_OFFSET);
+    }
+
+    private int getInt(int fieldOffset) throws IOException {
+        return in.readInt(nodeId * SERIALIZABLE_DATA_SIZE_IN_BYTES + fieldOffset);
+    }
+
+    private long getLong(int fieldOffset) throws IOException {
+        return in.readLong(nodeId * SERIALIZABLE_DATA_SIZE_IN_BYTES + fieldOffset);
+    }
+
+    private byte getByte(int fieldOffset) throws IOException {
+        return in.readByte(nodeId * SERIALIZABLE_DATA_SIZE_IN_BYTES + fieldOffset);
+    }
+
+    @Override
+    public int getDimensionId() throws IOException {
+        return getInt(DIMENSION_ID_OFFSET);
+    }
+
+    @Override
+    public long getDimensionValue() throws IOException {
+        return getLong(DIMENSION_VALUE_OFFSET);
+    }
+
+    @Override
+    public int getChildDimensionId() throws IOException {
+        if (firstChildId == INVALID_ID) {
+            return INVALID_ID;
+        } else {
+            return in.readInt(firstChildId * SERIALIZABLE_DATA_SIZE_IN_BYTES);
+        }
+    }
+
+    @Override
+    public int getStartDocId() throws IOException {
+        return getInt(START_DOC_ID_OFFSET);
+    }
+
+    @Override
+    public int getEndDocId() throws IOException {
+        return getInt(END_DOC_ID_OFFSET);
+    }
+
+    @Override
+    public int getAggregatedDocId() throws IOException {
+        return getInt(AGGREGATE_DOC_ID_OFFSET);
+    }
+
+    @Override
+    public int getNumChildren() throws IOException {
+        if (firstChildId == INVALID_ID) {
+            return 0;
+        } else {
+            return getInt(LAST_CHILD_ID_OFFSET) - firstChildId + 1;
+        }
+    }
+
+    @Override
+    public boolean isLeaf() {
+        return firstChildId == INVALID_ID;
+    }
+
+    @Override
+    public boolean isStarNode() throws IOException {
+        return getByte(IS_STAR_NODE_OFFSET) != 0;
+    }
+
+    @Override
+    public StarTreeNode getChildForDimensionValue(long dimensionValue) throws IOException {
+        // there will be no children for leaf nodes
+        if (isLeaf()) {
+            return null;
+        }
+
+        // Specialize star node for performance
+        if (dimensionValue == ALL) {
+            return handleStarNode();
+        }
+
+        return binarySearchChild(dimensionValue);
+    }
+
+    private FixedLengthStarTreeNode handleStarNode() throws IOException {
+        FixedLengthStarTreeNode firstNode = new FixedLengthStarTreeNode(in, firstChildId);
+        if (firstNode.getDimensionValue() == ALL) {
+            return firstNode;
+        } else {
+            return null;
+        }
+    }
+
+    private FixedLengthStarTreeNode binarySearchChild(long dimensionValue) throws IOException {
```

Comment:
Lets ensure to have couple of tests that children are written and read in sorted order.  ( we do sort the children during tests, but lets ensure to have tests )

Reply:
Major : 
How does it work when a level has null, -1 and star ?

Reply:
Introduced a new StarNodeType, which maintains if the node is star, null or normal nodes.
We sort on the StarNodeType first, and then on the dimension values for other nodes.

The order will be Star Node - Null Node (if present) - Normal Nodes (sorted on dimension values)